### PR TITLE
redpanda: convert post-install-upgrade-job.yaml to go

### PR DESCRIPTION
**Pull request by jan-g**
This includes a new `helmette.MergeTo`, as follows:

We are largely dealing with types from three domains: the input values, the desired k8s structures (both of these have relatively sophisticated typing; although these types might in some cases be roughly isomorphic but not identical) and the raw-json-structs of Helm's templates.

Because we're coming from a templated world, the use of `merge` is common to fold overrides onto defaults.

We provide a type-aware `MergeTo`, parameterised by the target type. This permits go to be written in a type-safe fashion, but still permits the kind of structure-merging common in code that begins life as the translation of a Helm template. Under the hood, it has to round-trip into `mergo`-compatible raw values, but we do our best to hide this from the caller.

### CODE REVIEW

File: charts/redpanda/post_install_upgrade_job.go
Comment (jan-g) [inline charts/redpanda/post_install_upgrade_job.go:89]: Question for reviewer: this merge is arguably the wrong way around, since the "defaults" will overrider the job-specific values. Given that this field was originally missing from values.go, I question whether this was actually being used in the wild.

I'm leaning towards fixing up the ordering of the MergeTo arguments here, but I'd like a second opinion before I make that change in this PR.

File: charts/redpanda/values.go
Comment (RafalKorepta) [reply charts/redpanda/values.go:490]: > do we have a way to audit the use of these potentially vestigial fields?

No, there is no way to check what values are used in the wild.

> which this PR currently doesn't fix - should it?

In the new commit that would have dedicated values input to test it out it can go in this PR or in new. I'm open to any outcome. 

>  I'm also not 100% convinced that additional security context settings can provide anything that this script could conceivably require in order to operate (outside of pathological scenarios).

I feel like access to `/tmp/` file system and network connectivity should be an requirement. Other security settings could be more restrictive.

File: charts/redpanda/post_install_upgrade_job.go
Comment (RafalKorepta) [inline charts/redpanda/post_install_upgrade_job.go:183]: This code transpiles to 

```
{{- $result = (concat (default (list ) $result) (list (default (dict ) $t).AsMap)) -}}
```

An my initial reaction would be how to test a case where `values.Tolerations` have nil map included in the list.

File: charts/redpanda/post_install_upgrade_job.go
Comment (RafalKorepta) [inline charts/redpanda/post_install_upgrade_job.go:87]: I just wonder why not:

```suggestion
							Resources: helmette.MergeTo[corev1.ResourceRequirements](resources),
```

File: charts/redpanda/post_install_upgrade_job.go
Comment (jan-g) [reply charts/redpanda/post_install_upgrade_job.go:48]: I'll fix all of these and put up a test case covering them.

File: charts/redpanda/post_install_upgrade_job.go
Comment (jan-g) [reply charts/redpanda/post_install_upgrade_job.go:87]: Yeah, given how trivial that function became once I had MergeTo I'll do this.

**Review by RafalKorepta — COMMENTED** [review]
I would like to approve this PR once you create new ci test case that would try to change PodSpec.Affinity and PodSpec.Tolerations and postInstallJobResources. As far as I saw the SecurityContext is correctly propagated within `34-security-contexts-novalues.yaml.golden` generated file.

File: charts/redpanda/post_install_upgrade_job.go
Comment (jan-g) [reply charts/redpanda/post_install_upgrade_job.go:183]: What would that mean? I think we should look to move towards a situation where there's a strong schema wrapped around `values` - Tolerations moving from `[]map[string]any` (ie, abrogating responsibility of checking the type to the user) to `[]corev1.Toleration`. Shifting toward the situation where illegal values are unrepresentable seems the right way to go.

File: charts/redpanda/post_install_upgrade_job.go
Comment (jan-g) [reply charts/redpanda/post_install_upgrade_job.go:87]: ... actually there's something amiss here. That refactor leaves the resulting template failing when values.postInstallJob.resources is empty.

File: charts/redpanda/post_install_upgrade_job.go
Comment (RafalKorepta) [reply charts/redpanda/post_install_upgrade_job.go:183]: I agree with you. Could we change the schema
https://github.com/redpanda-data/helm-charts/blob/9410e53301116e3aa5c6b6bc6cd49e51096ca86e/charts/redpanda/values.schema.json#L2487-L2492
to not allow nil object in the array?

File: charts/redpanda/post_install_upgrade_job.go
Comment (chrisseto) [inline charts/redpanda/post_install_upgrade_job.go:68]: Fun fact, printf has builtin support for truncating `%[50]s-post-install`. Either way works for me.

File: charts/redpanda/post_install_upgrade_job.go
Comment (chrisseto) [inline charts/redpanda/post_install_upgrade_job.go:188]: I have this solved in a commit I'll put up soon. I just added the corev1.ResourceRequirements to the values type.

File: charts/redpanda/post_install_upgrade_job.go
Comment (jan-g) [reply charts/redpanda/post_install_upgrade_job.go]: Fixed up in secnod commit.

File: charts/redpanda/post_install_upgrade_job.go
Comment (jan-g) [reply charts/redpanda/post_install_upgrade_job.go:68]: ah-ha, I'll fix that up.

File: pkg/gotohelm/helmette/sprig.go
Comment (jan-g) [reply pkg/gotohelm/helmette/sprig.go]: done anyway

File: charts/redpanda/post_install_upgrade_job.go
Comment (jan-g) [reply charts/redpanda/post_install_upgrade_job.go:48]: Change of plan: I've currently left this as-is. I think a general task to take an overview of overrides and ensure they're working correctly is better done over the setup as a whole rather than just here.

**Review by RafalKorepta — APPROVED** [review]

**Pull request merged**
